File: android/src/main/kotlin/com/descope/flutter/DescopePlugin.kt
package com.descope.flutter

import android.content.Context
import android.net.Uri
import android.os.Handler
import android.os.Looper
import androidx.browser.customtabs.CustomTabsIntent
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKeys
import com.descope.Descope
import com.descope.android.DescopeSystemInfo
import com.descope.internal.routes.getPackageOrigin
import com.descope.internal.routes.performAssertion
import com.descope.internal.routes.performNativeAuthorization
import com.descope.internal.routes.performRegister
import com.descope.sdk.DescopeLogger
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import org.json.JSONObject
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

/** DescopePlugin */
class DescopePlugin: FlutterPlugin, MethodCallHandler, ActivityAware {
  private var channel : MethodChannel? = null
  private var logChannel: MethodChannel? = null
  private var context: Context? = null
  private lateinit var storage: Store

  // MethodCallHandler
  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "getSystemInfo" -> getSystemInfo(result)
      "startFlow" -> startFlow(call, result)
      "oauthNative" -> oauthNative(call, result)
      "passkeySupported" -> isPasskeySupported(result)
      "passkeyOrigin" -> passkeyOrigin(result)
      "passkeyCreate" -> createOrUsePasskey(call, result, true)
      "passkeyAuthenticate" -> createOrUsePasskey(call, result, false)
      "loadItem" -> loadItem(call, result)
      "saveItem" -> saveItem(call, result)
      "removeItem" -> removeItem(call, result)
      else -> result.notImplemented()
    }
  }

  // General

  private fun getSystemInfo(res: Result) {
    val context = this.context ?: return res.error("NULLCONTEXT", "Context is null", null)
    val systemInfo = DescopeSystemInfo.getInstance(context)
    val info = mutableMapOf(
      "platformName" to "android",
      "platformVersion" to systemInfo.platformVersion,
    ).apply {
      systemInfo.appName?.let { put("appName", it) }
      systemInfo.appVersion?.let { put("appVersion", it) }
      systemInfo.device?.let { put("device", it) }
    }
    res.success(info)
  }

  // Flows - Deprecated

  private fun startFlow(call: MethodCall, res: Result) {
    val context = this.context ?: return res.error("NULLCONTEXT", "Context is null", null)
    val url = call.argument<String>("url") ?: return res.error("MISSINGARGS", "'url' is required for startFlow", null)
    try {
      val uri = Uri.parse(url)
      launchUri(context, uri)
      res.success(url)
    } catch (ignored: Exception) {
      res.error("INVALIDARGS", "url argument is invalid", null)
    }
  }

  // OAuth

  private fun oauthNative(call: MethodCall, res: Result) {
    val context = this.context ?: return res.error("NULLCONTEXT", "Context is null", null)
    val clientId = call.argument<String>("clientId") ?: return res.error("MISSINGARGS", "'clientId' is required for oauthNative", null)
    val nonce = call.argument<String>("nonce") ?: return res.error("MISSINGARGS", "'nonce' is required for oauthNative", null)
    val implicit = true // always true for Android
    GlobalScope.launch(Dispatchers.Main) {
      try {
        val identityToken = performNativeAuthorization(context, clientId, nonce, implicit)
        val json = JSONObject(mapOf("identityToken" to identityToken)).toString()
        res.success(json)
      } catch (e: Exception) {
        res.error("FAILED", e.message, null)
      }
    }
  }

  // Passkeys

  private fun isPasskeySupported(res: Result) {
    res.success(android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.P)
  }

  private fun passkeyOrigin(res: Result) {
    val context = this.context ?: return res.error("NULLCONTEXT", "Context is null", null)
    try {
      val origin = getPackageOrigin(context)
      res.success(origin)
    } catch (e: Exception) {
      res.error("FAILED", "Context is null", null)
    }
  }

  private fun createOrUsePasskey(call: MethodCall, res: Result, create: Boolean) {
    val context = this.context ?: return res.error("NULLCONTEXT", "Context is null", null)
    val options = call.argument<String>("options") ?: return res.error("MISSINGARGS", "'options' is required for passkey functions", null)
    GlobalScope.launch(Dispatchers.Main) {
      try {
        val response =
          if (create) performRegister(context, options)
          else performAssertion(context, options)
        res.success(response)
      } catch (e: Exception) {
        res.error("FAILED", e.message, null)
      }
    }
  }

  // Storage

  @Suppress("UNUSED_PARAMETER")
  private fun initStorageIfNeeded(call: MethodCall, key: String, res: Result): Boolean {
    if (this::storage.isInitialized) return false

    val context = this.context
    if (context == null) {
      res.error("NULLCONTEXT", "Context is null", null)
      return true
    }

    storage = createEncryptedStore(context, key)
    return false
  }

  private fun loadItem(call: MethodCall, res: Result) {
    val key = keyFromCall(call, res) ?: return
    if (initStorageIfNeeded(call, key, res)) return
    val value = storage.loadItem(key)
    res.success(value)
  }

  private fun saveItem(call: MethodCall, res: Result) {
    val key = keyFromCall(call, res) ?: return
    if (initStorageIfNeeded(call, key, res)) return
    val data = dataFromCall(call, res) ?: return
    storage.saveItem(key, data)
    res.success(key)
  }

  private fun removeItem(call: MethodCall, res: Result) {
    val key = keyFromCall(call, res) ?: return
    if (initStorageIfNeeded(call, key, res)) return
    storage.removeItem(key)
    res.success(key)
  }

  private fun keyFromCall(call: MethodCall, res: Result) = stringFromCall("key", call, res)

  private fun dataFromCall(call: MethodCall, res: Result) = stringFromCall("data", call, res)

  private fun stringFromCall(key: String, call: MethodCall, res: Result): String? {
    val data = call.argument<String>(key)
    if (data == null) {
      res.error("MISSINGARGS", "'$key' argument is required", null)
    }
    return data
  }

  // FlutterPlugin

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "descope_flutter/methods")
    channel?.setMethodCallHandler(this)

    // Set up the log channel with a handler for logger configuration from Flutter
    val logsChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "descope_flutter/logs")
    logChannel = logsChannel
    val applicationContext = flutterPluginBinding.applicationContext

    logsChannel.setMethodCallHandler { call, result ->
      if (call.method == "configure") {
        val levelString = call.argument<String>("level")
        val unsafe = call.argument<Boolean>("unsafe")

        if (levelString == null || unsafe == null) {
          result.error("INVALID_ARGS", "Missing level or unsafe arguments", null)
          return@setMethodCallHandler
        }

        val level = when (levelString) {
          "error" -> DescopeLogger.Level.Error
          "info" -> DescopeLogger.Level.Info
          else -> DescopeLogger.Level.Debug
        }

        // Initialize SDK with the logger configured from Flutter
        Descope.setup(applicationContext, projectId = "") {
          logger = FlutterDescopeLogger(logsChannel, level, unsafe)
        }

        result.success(null)
      } else {
        result.notImplemented()
      }
    }

    flutterPluginBinding.platformViewRegistry.registerViewFactory(
      "descope_flutter/descope_flow_view",
      DescopeFlowViewFactory(flutterPluginBinding.binaryMessenger)
    )
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel?.setMethodCallHandler(null)
    channel = null
    logChannel?.setMethodCallHandler(null)
    logChannel = null
  }

  // ActivityAware

  override fun onAttachedToActivity(activityPluginBinding: ActivityPluginBinding) {
    context = activityPluginBinding.activity
  }

  override fun onDetachedFromActivityForConfigChanges() {
    context = null
  }

  override fun onReattachedToActivityForConfigChanges(activityPluginBinding: ActivityPluginBinding) {
    context = activityPluginBinding.activity
  }

  override fun onDetachedFromActivity() {
    context = null
  }
}

private fun launchUri(context: Context, uri: Uri) {
  val customTabsIntent = CustomTabsIntent.Builder()
    .setUrlBarHidingEnabled(true)
    .setShowTitle(true)
    .setShareState(CustomTabsIntent.SHARE_STATE_OFF)
    .build()
  customTabsIntent.launchUrl(context, uri)
}

private class EncryptedStorage(context: Context, name: String): Store {
  private val masterKey = MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC)
  private val sharedPreferences = EncryptedSharedPreferences.create(
    name,
    masterKey,
    context,
    EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
    EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
  )

  override fun loadItem(key: String): String? = sharedPreferences.getString(key, null)

  override fun saveItem(key: String, data: String) = sharedPreferences.edit()
    .putString(key, data)
    .apply()

  override fun removeItem(key: String) = sharedPreferences.edit()
    .remove(key)
    .apply()
}

/**
 * A helper interface that takes care of the actual storage of session data.
 *
 * The default function implementations in this interface do nothing or return `null`.
 */
interface Store {
  fun saveItem(key: String, data: String) {}

  fun loadItem(key: String): String? = null

  fun removeItem(key: String) {}

  companion object {
    /** A store that does nothing */
    val none = object : Store {}
  }
}

private fun createEncryptedStore(context: Context, projectId: String): Store {
  try {
    return EncryptedStorage(context, projectId)
  } catch (e: Exception) {
    try {
      // encrypted storage key unusable
      context.deleteSharedPreferences(projectId)
      return EncryptedStorage(context, projectId)
    } catch (e: Exception) {
      // unable to initialize encrypted storage
      return Store.none
    }
  }
}

// Logger

/**
 * A DescopeLogger subclass that forwards all logs to Flutter via a MethodChannel.
 * This logger mirrors the level and unsafe settings from the Flutter layer.
 */
private class FlutterDescopeLogger(private val channel: MethodChannel, level: Level, unsafe: Boolean) : DescopeLogger(level, unsafe) {
    private val handler = Handler(Looper.getMainLooper())

    override fun output(level: Level, message: String, values: List<Any>) {
        val levelString = when (level) {
            Level.Error -> "error"
            Level.Info -> "info"
            Level.Debug -> "debug"
        }

        val valuesArray = values.map { it.toString() }

        handler.post {
            channel.invokeMethod("log", mapOf(
                "level" to levelString,
                "message" to message,
                "values" to valuesArray,
            ))
        }
    }
}

File: ios/descope/Sources/descope/DescopePlugin.swift
import AuthenticationServices
import Flutter

private let redirectScheme = "descopeauth"
private let redirectURL = "\(redirectScheme)://flow"

public class DescopePlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    
    private let defaultContextProvider = FlowsV0DefaultContextProvider()
    private let keychainStore = KeychainStore()
    private var eventSink: FlutterEventSink?
    private var sessions: [ASWebAuthenticationSession] = []
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(name: "descope_flutter/methods", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "descope_flutter/events", binaryMessenger: registrar.messenger())
        let logChannel = FlutterMethodChannel(name: "descope_flutter/logs", binaryMessenger: registrar.messenger())
        
        let instance = DescopePlugin()

        // Set up log channel handler for logger configuration from Flutter
        logChannel.setMethodCallHandler { call, result in
            if call.method == "configure" {
                guard let args = call.arguments as? [String: Any],
                      let levelString = args["level"] as? String,
                      let unsafe = args["unsafe"] as? Bool else {
                    result(FlutterError(code: "INVALID_ARGS", message: "Missing level or unsafe arguments", details: nil))
                    return
                }

                let level: DescopeLogger.Level
                switch levelString {
                case "error":
                    level = .error
                case "info":
                    level = .info
                default:
                    level = .debug
                }

                // Initialize SDK with the logger configured from Flutter
                // Descope.setup is @MainActor so we ensure we're on the main thread
                DispatchQueue.main.async {
                    Descope.setup(projectId: "") { config in
                        config.logger = FlutterDescopeLogger(channel: logChannel, level: level, unsafe: unsafe)
                    }
                }

                result(nil)
            } else {
                result(FlutterMethodNotImplemented)
            }
        }

        eventChannel.setStreamHandler(instance)
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        let factory = FlutterDescopeFlowViewFactory(messenger: registrar.messenger())
        registrar.register(factory, withId: "descope_flutter/descope_flow_view")
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getSystemInfo":
            getSystemInfo(result: result)
        case "startFlow":
            startFlow(call: call, result: result)
        case "oauthNative":
            oauthNative(call: call, result: result)
        case "passkeySupported":
            passkeySupported(result: result)
        case "passkeyOrigin":
            result("") // No need for passkey origin on iOS
        case "passkeyCreate":
            createPasskey(call: call, result: result)
        case "passkeyAuthenticate":
            usePasskey(call: call, result: result)
        case "loadItem":
            loadItem(call: call, result: result)
        case "saveItem":
            saveItem(call: call, result: result)
        case "removeItem":
            removeItem(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // General

    private func getSystemInfo(result: @escaping FlutterResult) {
        result([
            "platformName": SystemInfo.osName,
            "platformVersion": SystemInfo.osVersion,
            "appName": SystemInfo.appName,
            "appVersion": SystemInfo.appVersion,
            "device": SystemInfo.device,
        ])
    }
    
    // Flows
    
    private func startFlow(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any], let urlString = args["url"] as? String else { return result(FlutterError(code: "MISSINGARGS", message: "'url' is required for startFlow", details: nil)) }
        DispatchQueue.main.async { [self] in
            startFlow(urlString)
        }
        result(urlString)
    }
    
    // OAuth

    private func oauthNative(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any] else { return result(FlutterError(code: "MISSINGARGS", message: "Unexpected empty arguments in oauthNative", details: nil)) }
        guard let nonce = args["nonce"] as? String else { return result(FlutterError(code: "MISSINGARGS", message: "'nonce' is required for oauthNative", details: nil)) }
        guard let implicit = args["implicit"] as? Bool else { return result(FlutterError(code: "MISSINGARGS", message: "'implicit' is required for oauthNative", details: nil)) }

        Task { @MainActor in
            do {
                let (authorizationCode, identityToken, user) = try await OAuth.performNativeAuthentication(nonce: nonce, implicit: implicit, logger: nil)

                let values = [
                    "authorizationCode": authorizationCode,
                    "identityToken": identityToken,
                    "user": user,
                ]

                let json = try JSONSerialization.data(withJSONObject: values, options: [])
                guard let response = String(bytes: json, encoding: .utf8) else { return result(FlutterError(code: "FAILED", message: "Response encoding failed", details: nil)) }

                return result(response)
            } catch let error as DescopeError {
                return result(FlutterError(from: error))
            }
        }
    }
    
    // Passkeys
    
    private func passkeySupported(result: @escaping FlutterResult) {
        if #available(iOS 15, *) {
            result(true)
        }
        result(false)
    }
    
    private func createPasskey(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard #available(iOS 15, *) else { return result(FlutterError(code: "OSVERSION", message: "Passkeys require iOS 15 and above", details: nil)) }
        guard let args = call.arguments as? [String: Any] else { return result(FlutterError(code: "MISSINGARGS", message: "Unexpected empty arguments in passkeyCreate", details: nil)) }
        guard let options = args["options"] as? String else { return result(FlutterError(code: "MISSINGARGS", message: "'options' is required for passkeyCreate", details: nil)) }
        Task { @MainActor in
            do {
                let response = try await Passkey.performRegister(options: options, logger: nil)
                result(response)
            } catch let error as DescopeError {
                return result(FlutterError(from: error))
            }
        }
    }
    
    private func usePasskey(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard #available(iOS 15, *) else { return result(FlutterError(code: "OSVERSION", message: "Passkeys require iOS 15 and above", details: nil)) }
        guard let args = call.arguments as? [String: Any] else { return result(FlutterError(code: "MISSINGARGS", message: "Unexpected empty arguments in passkeyAuthenticate", details: nil)) }
        guard let options = args["options"] as? String else { return result(FlutterError(code: "MISSINGARGS", message: "'options' is required for passkeyAuthenticate", details: nil)) }
        Task { @MainActor in
            do {
                let response = try await Passkey.performAssertion(options: options, logger: nil)
                result(response)
            } catch let error as DescopeError {
                return result(FlutterError(from: error))
            }
        }
    }
    
    // Storage
    
    private func loadItem(call: FlutterMethodCall, result: FlutterResult) {
        guard let args = call.arguments as? [String: Any], let key = args["key"] as? String else { return result(FlutterError(code: "MISSINARGS", message: "'key' is required for loadItem", details: nil)) }
        guard let data = keychainStore.loadItem(key: key) else { return result(nil) }
        let value = String(bytes: data, encoding: .utf8)
        result(value)
    }
    
    private func saveItem(call: FlutterMethodCall, result: FlutterResult) {
        guard let args = call.arguments as? [String: Any], let key = args["key"] as? String, let value = args["data"] as? String else { return result(FlutterError(code: "MISSINARGS", message: "'key' and 'data' are required for saveItem", details: nil)) }
        keychainStore.saveItem(key: key, data: Data(value.utf8))
        result(key)
    }
    
    private func removeItem(call: FlutterMethodCall, result: FlutterResult) {
        guard let args = call.arguments as? [String: Any], let key = args["key"] as? String else { return result(FlutterError(code: "MISSINARGS", message: "'key' is required for removeItem", details: nil)) }
        keychainStore.removeItem(key: key)
        result(key)
    }
    
    // FlutterStreamHandler
    
    public func onListen(withArguments arguments: Any?, eventSink: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = eventSink
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
    
    // Internal Flows

    @MainActor
    private func startFlow(_ urlString: String) {
        startFlow(urlString, attempts: 5)
    }

    @MainActor
    private func startFlow(_ urlString: String, attempts: Int) {
        if attempts > 0, defaultContextProvider.findKeyWindow() == nil {
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.2) { [self] in
                startFlow(urlString, attempts: attempts - 1)
            }
            return
        }
        
        guard let url = URL(string: urlString) else { return }
        let session = ASWebAuthenticationSession(url: url, callbackURLScheme: redirectScheme) { [self] callbackURL, error in
            if let error {
                switch error {
                case ASWebAuthenticationSessionError.canceledLogin:
                    respond(with: "canceled")
                    return
                case ASWebAuthenticationSessionError.presentationContextInvalid,
                    ASWebAuthenticationSessionError.presentationContextNotProvided:
                    // not handled for now
                    fallthrough
                default:
                    respond(with: "")
                    return
                }
            }
            respond(with: callbackURL?.absoluteString ?? "")
        }
        session.prefersEphemeralWebBrowserSession = true
        session.presentationContextProvider = defaultContextProvider
        sessions += [session]
        session.start()
    }
    
    @MainActor
    private func respond(with response: String) {
        eventSink?(response)
        for session in sessions {
            session.cancel()
        }
        sessions = []
    }
}

private class FlowsV0DefaultContextProvider: NSObject, ASWebAuthenticationPresentationContextProviding, ASAuthorizationControllerPresentationContextProviding {
    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        return findKeyWindow() ?? ASPresentationAnchor()
    }

    func presentationAnchor(for controller: ASAuthorizationController) -> ASPresentationAnchor {
        return findKeyWindow() ?? ASPresentationAnchor()
    }

    func findKeyWindow() -> UIWindow? {
        let scene = UIApplication.shared.connectedScenes
            .filter { $0.activationState == .foregroundActive }
            .compactMap { $0 as? UIWindowScene }
            .first

        let keyWindow = scene?.windows
            .first { $0.isKeyWindow }

        return keyWindow
    }
}

private class KeychainStore {
    public func loadItem(key: String) -> Data? {
        var query = queryForItem(key: key)
        query[kSecReturnData as String] = true
        query[kSecMatchLimit as String] = kSecMatchLimitOne
        
        var value: AnyObject?
        SecItemCopyMatching(query as CFDictionary, &value)
        return value as? Data
    }
    
    public func saveItem(key: String, data: Data) {
        let values: [String: Any] = [
            kSecValueData as String: data,
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlock,
        ]

        let query = queryForItem(key: key)
        let result = SecItemCopyMatching(query as CFDictionary, nil)
        if result == errSecSuccess {
            SecItemUpdate(query as CFDictionary, values as CFDictionary)
        } else if result == errSecItemNotFound {
            let merged = query.merging(values, uniquingKeysWith: { $1 })
            SecItemAdd(merged as CFDictionary, nil)
        }
    }
    
    public func removeItem(key: String) {
        let query = queryForItem(key: key)
        SecItemDelete(query as CFDictionary)
    }
    
    private func queryForItem(key: String) -> [String: Any] {
        return [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: "com.descope.Flutter",
            kSecAttrLabel as String: "DescopeSession",
            kSecAttrAccount as String: key,
        ]
    }
}

// Passkey & OAuth Errors

extension FlutterError {
    convenience init(from error: DescopeError) {
        let code: String
        switch error {
        case DescopeError.passkeyCancelled, DescopeError.oauthNativeCancelled, DescopeError.flowCancelled:
            code = "CANCELLED"
        default:
            code = "FAILED"
        }

        let message: String
        if let msg = error.message {
            message = "\(error.desc): \(msg)"
        } else {
            message = error.desc
        }

        self.init(code: code, message: message, details: nil)
    }
}

/// A DescopeLogger subclass that forwards all logs to Flutter via a MethodChannel.
/// This logger is configured with the level and unsafe settings from the Flutter layer.
private class FlutterDescopeLogger: DescopeLogger {
    private let channel: FlutterMethodChannel

    init(channel: FlutterMethodChannel, level: Level, unsafe: Bool) {
        self.channel = channel
        super.init(level: level, unsafe: unsafe)
    }

    override func output(level: Level, message: String, unsafe values: [Any]) {
        let levelString: String
        switch level {
        case .error:
            levelString = "error"
        case .info:
            levelString = "info"
        case .debug:
            levelString = "debug"
        }

        let valuesArray = values.map { String(describing: $0) }

        DispatchQueue.main.async {
            self.channel.invokeMethod("log", arguments: [
                "level": levelString,
                "message": message,
                "values": valuesArray,
            ])
        }
    }
}
